### PR TITLE
fix(typings): fix typings for error prone operators like concatAll

### LINKED ISSUE #2658: Exhaust causes type issues
**RxJS version: 5.0.1**

**Code to reproduce:**
.exhaust()

**Expected behavior:**
Flattens an Observable-of-Observables by dropping the next inner Observables
while the current inner is still executing.

**Actual behavior:**
Does not flatten the Observable.

**Additional information:**
There was a change in this commit 
[7b23e88](https://github.com/ReactiveX/rxjs/commit/96377faa58ce1f23ca8422dcf48028b2912d2467#diff-20d891a331152011cdf8fffd501873f1),
which updates the type of exhaust from **SwitchFirstSignature<T>** to **typeof exhaust**.
The first one is of type Generic, while the other one's type is Observable.
This will cause Typescript issues such as "error TS2322: Type 'Observable<any>' is not assignable to type..."

### PR DESCRIPTION
**Pull request by david-driscoll**
**Description:**
fixes concatAll, combineAll, exhaust, mergeAll, switch, zipAll.  Also made a few improvements to
how hot and cold observables are typed.  A few compiler errors were fixed due to this change.

**Related issue (if exists):**
fixes #2658 #2493 #2551

### CODE REVIEW

File: spec/helpers/marble-testing.ts [hunk 1]
```diff
@@ -1,45 +1,49 @@
...
-  }
-  return global.rxTestScheduler.expectSubscriptions.apply(global.rxTestScheduler, arguments);
-}
-
-export function time(marbles: string): number {
-  if (!global.rxTestScheduler) {
-    throw 'tried to use time() in async test';
-  }
-  return global.rxTestScheduler.createTime.apply(global.rxTestScheduler, arguments);
+import {Observable} from '../../dist/cjs/Observable';
```
Comment (david-driscoll) [inline spec/helpers/marble-testing.ts]: I have no idea why this file diffed so badly... but this adds two overloads for `hot` and `cold` to help with type inference.

Comment (kwonoj) [inline spec/helpers/marble-testing.ts]: line endings?

Comment (david-driscoll) [inline spec/helpers/marble-testing.ts]: just pushed a change to line endings and it didn't change anything :(

**Comment by david-driscoll** [comment]
This also has changes that help with #1633

File: spec/helpers/marble-testing.ts [hunk 2]
```diff
@@ -1,45 +1,49 @@
...
+import {SubscriptionLog} from '../../dist/cjs/testing/SubscriptionLog';
+import {ColdObservable} from '../../dist/cjs/testing/ColdObservable';
+import {HotObservable} from '../../dist/cjs/testing/HotObservable';
+import {TestScheduler, observableToBeFn, subscriptionLogsToBeFn} from '../../dist/cjs/testing/TestScheduler';
+
+declare const global: any;
+
+export const rxTestScheduler: TestScheduler = global.rxTestScheduler;
+
+export function hot(marbles: string, values?: void, error?: any): HotObservable<string>;
```
Comment (kwonoj) [inline spec/helpers/marble-testing.ts]: uh, `values` is void type?

Comment (david-driscoll) [inline spec/helpers/marble-testing.ts]: Yeah, we've used it before for things like `groupBy` and `bindNodeCallback`

Comment (david-driscoll) [inline spec/helpers/marble-testing.ts]: Rather values it void by default (if we don't give it a value), if we do give it a value we can then infer it's type.

**Comment by david-driscoll** [comment]
any input @kwonoj @benlesh ? 🕺 💃

File: spec/helpers/marble-testing.ts (hunk 1 above)
Comment (benlesh) [inline spec/helpers/marble-testing.ts]: Hmm... it's hard to review like this.

Is this maybe a thing from working on a windows machine?

Comment (david-driscoll) [inline spec/helpers/marble-testing.ts]: I had reset line endings on the file and it didn't change the diff.  It diffs just fine for me locally too :(

Comment (benlesh) [inline spec/helpers/marble-testing.ts]: Can you comment on what's changed in this file?

Comment (david-driscoll) [inline spec/helpers/marble-testing.ts]: There are two new overloads for `hot` and `cold`.

```typescript
export function hot(marbles: string, values?: void, error?: any): HotObservable<string>;
export function hot<V>(marbles: string, values?: { [index: string]: V; }, error?: any): HotObservable<V>;


export function cold(marbles: string, values?: void, error?: any): ColdObservable<string>;
export function cold<V>(marbles: string, values?: { [index: string]: V; }, error?: any): ColdObservable<V>;
```

The first overload, defaults the returned `HotObservable` to be a type of string (ie `ColdObservable<string>`)
The second overload, allows for type inference of the output observable, so that it is typed as a union of the input values.

Given a value  `let values = { a: 1, b: 2, c: 3 }` then the type of `hot('a---b---c---|', values)` will then `HotObservable<number>`;

Given a more complex case (stolen from `mergeAll-spec.ts`)
```typescript
it('should merge two cold Observables at a time with parameter concurrency=2', () => {
    const x = cold(     'a---b---c---|        ');
    const xsubs =     '  ^           !        ';
    const y = cold(        'd---e---f---|     ');
    const ysubs =     '     ^           !     ';
    const z = cold(                 '--g---h-|');
    const zsubs =     '              ^       !';
    const e1 =    hot('--x--y--z--|           ', { x, y, z });
    const e1subs =    '^                     !';
    const expected =  '--a--db--ec--f--g---h-|';

    expectObservable(e1.mergeAll(2)).toBe(expected);
    expectSubscriptions(x.subscriptions).toBe(xsubs);
    expectSubscriptions(y.subscriptions).toBe(ysubs);
    expectSubscriptions(z.subscriptions).toBe(zsubs);
    expectSubscriptions(e1.subscriptions).toBe(e1subs);
  });
```

Given that `x`, `y`, and `z` are all `ColdObservable<string>` the type of `e1` then becomes `HotObservable<ColdObservable<string>>`.  This then gives us proper type information in unit tests and allows removing of several corner cases where we explicitly cast to any to work around the problem.

Comment (benlesh) [inline spec/helpers/marble-testing.ts]: I see ... looks good.
